src/shared/components/case-editor/services/event-completion-state-machine.service.ts

```diff
@@ -0,0 +1,175 @@
+
+const EVENT_COMPLETION_STATE_MACHINE = 'EVENT COMPLETION STATE MACHINE';
+
+@Injectable()
+export class EventCompletionStateMachineService {
+  public stateCheckTasksCanBeCompleted: State;
+  public stateCompleteEventAndTask: State;
+  public stateCancelEvent: State;
+  public stateCompleteEventNotTask: State;
+  public stateTaskCompletedOrCancelled: State;
+  public stateTaskAssignedToAnotherUser: State;
+  public stateTaskReassignToUser: State;
+  public stateTaskAssignToUser: State;
+  public stateTaskUnassigned: State;
+  public stateFinal: State;
+
+  public initialiseStateMachine(context: EventCompletionStateMachineContext): StateMachine {
+    return new StateMachine(EVENT_COMPLETION_STATE_MACHINE, context);
+  }
+
+  public startStateMachine(stateMachine: StateMachine): void {
+    stateMachine.start(this.stateCheckTasksCanBeCompleted);
+  }
+
+  public createStates(stateMachine: StateMachine): void {
+    // Initial state
+    this.stateCheckTasksCanBeCompleted = stateMachine.createState(
+      EventCompletionStates.CHECK_TASKS_CAN_BE_COMPLETED,
+      false,
+      this.entryActionForStateCheckTasksCanBeCompleted
+    )
+
+    this.stateCompleteEventAndTask = stateMachine.createState(
+      EventCompletionStates.COMPLETE_EVENT_AND_TASK,
+      false,
+      this.entryActionForStateCompleteEventAndTask
+    )
+
+    this.stateTaskCompletedOrCancelled = stateMachine.createState(
+      EventCompletionStates.TASK_COMPLETED_OR_CANCELLED,
+      false,
+      this.entryActionForStateTaskCompletedOrCancelled
+    )
+
+    this.stateTaskAssignedToAnotherUser = stateMachine.createState(
+      EventCompletionStates.TASK_ASSIGNED_TO_ANOTHER_USER,
+      false,
+      this.entryActionForStateTaskAssignedToAnotherUser
+    )
+
+    this.stateTaskUnassigned = stateMachine.createState(
+      EventCompletionStates.TASK_UNASSIGNED,
+      false,
+      this.entryActionForStateTaskUnassigned
+    )
+
+    // Create final state, the second param isComplete is set to true to make sure state machine finished running
+    this.stateFinal = stateMachine.createState(
+      EventCompletionStates.FINAL,
+      true,
+      this.entryActionForStateFinal
+    );
+  }
+
+  public addTransitions(): void {
+    // Initial transition
+    this.addTransitionsForStateCheckTasksCanBeCompleted();
+    this.addTransitionsForStateCompleteEventAndTask();
+    this.addTransitionsForStateTaskAssignedToAnotherUser();
+    this.addTransitionsForStateTaskUnassigned();
+  }
+
+  public entryActionForStateCheckTasksCanBeCompleted(state: State, context: EventCompletionStateMachineContext): void {
+    context.workAllocationService.getTasksByCaseIdAndEventId(context.eventId, context.caseId).subscribe(payload => {
+      const taskPayLoad = <TaskPayload>payload;
+      if (taskPayLoad.task_required_for_event) {
+        const task = taskPayLoad.tasks.find(x => x.id === context.task.id);
+        if (task) {
+          if (!task.assignee && task.task_state === 'unassigned') {
```

Comment:
Can we not use the `enums` here?

Reply:
There are few more tickets to work on, and I will sort this out when implementing them.